text/0000-secret-types.md

```diff
@@ -0,0 +1,206 @@
+
+Applications deal with sensitive data all the time to varying degrees of success. Sensitive data is not limited to information like cryptographic codes and keys, but could also be data like passwords or PII such as social security numbers. Accidental secret leakage is a challenge for both programmers, who might mix secret and public data inadvertently, and compilers, which might use optimizations that reveal secrets via side-channels.
+
+Writing cryptographic and other security critical code in high-level languages like Rust is attractive for numerous reasons. High level languages are generally more readable and accessible to developers and reviewers, leading to higher quality, more secure code. It also allows the integration of cryptographic code with the rest of an application without requiring the use of any FFI. We are also typically motivated to have a reference implementation for algorithms that is portable to architectures that may not be supported by highly optimized assembly implementations. However, writing data invariant code in high level languages is difficult due to compiler optimizations. For this reason, having compiler support for a data type that is resistant to timing side channel attacks is desirable.
+
+Timing side channel attacks are a particular threat in a post spectre world [1]. Side channels are primarily used to attack secrets that are
+long lived
+ - Extremely valuable if compromised
+ - Each bit compromised provides incremental value
+ - Confidentiality of compromise is desirable
+Therefore, it’s important to use data-invariant programming for secrets. For this reason, secret types would only allow data invariant operations at all levels of compilation.
+
+Additionally, secret types serve as an indicator to programmers that this information should be treated with care, and not mixed with non-secret data, an invariant that would be enforced by the type system.
+
+[1] *Cryptographic Software in a Post-Spectre World.* Chandler Carruth. RWC 2020. Recording forthcoming.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Secret integer types are a type of restricted integer primitive. In particular, non-constant time operations like division are prohibited. Printing of secret integers directly is also prohibited--they must first be declassified into non-secret integers.
+
+These integers are intended to form a basis for the creation of other secret types, which can be built on top of them, such as secret strings for storing passwords in them.
+
+Comparison of secret integer types is allowed, but the algorithm is independent of the secrets. Comparison of secret types must return a secret_bool, which cannot be branched on.
+
+Example: comparison
+```
+let x : secret_i8 = 6;
+Let y : secret_i8 = 10;
+if (x < y) { // compiler error: cannot branch on secret bool
+  ...
+}
+```
+Example: declassification
+```
+let x : secret_i8 = 6;
+let y : secret_i8 = 10;
+println!((x ^ y).declassify())
```

Comment:
What does `declassify()` _do_?

Reply:
Explicitly copy the secret type to a public type. As an example, think of a ciphertext, which depends on a secret key and a secret message and is thus secret, being declassified before being sent over a channel.